File: examples/flask-middleware/strongdm_auth.py
"""
StrongDM ID Authentication Middleware for Flask

This middleware validates JWT tokens issued by StrongDM ID (https://id.strongdm.ai).
It supports:
- JWT signature verification using JWKS
- Token introspection fallback
- Scope-based access control
- DPoP token validation (optional)
"""

import time
import hashlib
import base64
import json
from functools import wraps
from typing import Optional, Callable

import jwt
import requests
from flask import Flask, request, g, jsonify, Response
from cachetools import TTLCache
from jwt import PyJWKClient, PyJWKClientError


class StrongDMAuthError(Exception):
    """Base exception for StrongDM auth errors."""
    def __init__(self, message: str, status_code: int = 401):
        self.message = message
        self.status_code = status_code
        super().__init__(message)


class StrongDMAuth:
    """
    StrongDM ID authentication middleware for Flask.

    Usage:
        app = Flask(__name__)
        auth = StrongDMAuth(app)

        @app.route('/protected')
        @auth.require_auth()
        def protected():
            return f"Hello, {g.token_claims['sub']}"

        @app.route('/admin')
        @auth.require_scope('admin')
        def admin():
            return "Admin area"
    """

    DEFAULT_ISSUER = "https://id.strongdm.ai"

    def __init__(
        self,
        app: Optional[Flask] = None,
        issuer: str = DEFAULT_ISSUER,
        audience: Optional[str] = None,
        jwks_cache_ttl: int = 900,  # 15 minutes
        introspection_enabled: bool = False,
        client_id: Optional[str] = None,
        client_secret: Optional[str] = None,
    ):
        """
        Initialize the StrongDM auth middleware.

        Args:
            app: Flask application instance
            issuer: Token issuer URL (default: https://id.strongdm.ai)
            audience: Expected audience claim (optional)
            jwks_cache_ttl: JWKS cache TTL in seconds (default: 15 minutes)
            introspection_enabled: Enable token introspection fallback
            client_id: Client ID for introspection (required if introspection enabled)
            client_secret: Client secret for introspection (required if introspection enabled)
        """
        self.issuer = issuer.rstrip('/')
        self.audience = audience
        self.jwks_cache_ttl = jwks_cache_ttl
        self.introspection_enabled = introspection_enabled
        self.client_id = client_id
        self.client_secret = client_secret

        # JWKS clients keyed by issuer URL
        self._jwks_clients: dict[str, PyJWKClient] = {}

        # Cache for introspection results
        self._introspection_cache = TTLCache(maxsize=1000, ttl=60)

        # Discovery document cache
        self._discovery: Optional[dict] = None
        self._discovery_fetched_at: float = 0

        if app is not None:
            self.init_app(app)

    def init_app(self, app: Flask) -> None:
        """Initialize the middleware with a Flask app."""
        app.config.setdefault('STRONGDM_ISSUER', self.issuer)
        app.config.setdefault('STRONGDM_AUDIENCE', self.audience)

        # Register error handler
        @app.errorhandler(StrongDMAuthError)
        def handle_auth_error(error: StrongDMAuthError) -> tuple[Response, int]:
            return jsonify({"error": error.message}), error.status_code

    def _validate_issuer(self, token_issuer: str) -> None:
        """Validate that the token issuer is the base or a realm-qualified issuer."""
        if token_issuer == self.issuer:
            return
        if token_issuer.startswith(self.issuer + '/realms/'):
            return
        raise StrongDMAuthError("Invalid token issuer")

    def _get_jwks_client_for_issuer(self, issuer: str) -> PyJWKClient:
        """Get or create a JWKS client for the given issuer."""
        if issuer not in self._jwks_clients:
            jwks_uri = issuer.rstrip('/') + '/jwks'
            self._jwks_clients[issuer] = PyJWKClient(
                jwks_uri,
                cache_jwk_set=True,
                lifespan=self.jwks_cache_ttl
            )
        return self._jwks_clients[issuer]

    def _get_discovery(self) -> dict:
        """Fetch and cache the OIDC discovery document."""
        now = time.time()
        if self._discovery is None or (now - self._discovery_fetched_at) > 3600:
            discovery_url = f"{self.issuer}/.well-known/openid-configuration"
            resp = requests.get(discovery_url, timeout=10)
            resp.raise_for_status()
            self._discovery = resp.json()
            self._discovery_fetched_at = now
        return self._discovery

    def _extract_token(self) -> tuple[str, str]:
        """
        Extract the token from the Authorization header.

        Returns:
            Tuple of (token_type, token_value)
        """
        auth_header = request.headers.get('Authorization', '')

        if not auth_header:
            raise StrongDMAuthError("Missing Authorization header")

        parts = auth_header.split(' ', 1)
        if len(parts) != 2:
            raise StrongDMAuthError("Invalid Authorization header format")

        token_type, token_value = parts
        token_type = token_type.lower()

        if token_type not in ('bearer', 'dpop'):
            raise StrongDMAuthError(f"Unsupported token type: {token_type}")

        return token_type, token_value

    @staticmethod
    def _jwk_thumbprint(jwk: dict) -> str:
        """Compute RFC 7638 SHA-256 JWK thumbprint."""
        kty = jwk.get("kty")
        if kty == "RSA":
            canonical_members = {"e": jwk.get("e"), "kty": kty, "n": jwk.get("n")}
        elif kty == "EC":
            canonical_members = {"crv": jwk.get("crv"), "kty": kty, "x": jwk.get("x"), "y": jwk.get("y")}
        elif kty == "OKP":
            canonical_members = {"crv": jwk.get("crv"), "kty": kty, "x": jwk.get("x")}
        elif kty == "oct":
            canonical_members = {"k": jwk.get("k"), "kty": kty}
        else:
            raise StrongDMAuthError("Invalid DPoP proof: unsupported JWK kty")

        if any(v is None for v in canonical_members.values()):
            raise StrongDMAuthError("Invalid DPoP proof: incomplete JWK for thumbprint")

        canonical = json.dumps(canonical_members, separators=(",", ":"), sort_keys=True).encode()
        digest = hashlib.sha256(canonical).digest()
        return base64.urlsafe_b64encode(digest).rstrip(b"=").decode()

    def _verify_dpop_proof(self, dpop_proof: str, access_token: str) -> tuple[dict, str]:
        """
        Verify a DPoP proof JWT.

        Args:
            dpop_proof: The DPoP proof JWT from the DPoP header
            access_token: The access token being used

        Returns:
            The decoded DPoP proof claims
        """
        # Decode header without verification to get the JWK
        unverified_header = jwt.get_unverified_header(dpop_proof)

        if unverified_header.get('typ') != 'dpop+jwt':
            raise StrongDMAuthError("Invalid DPoP proof: wrong typ")

        jwk = unverified_header.get('jwk')
        if not jwk:
            raise StrongDMAuthError("Invalid DPoP proof: missing jwk")

        proof_jkt = self._jwk_thumbprint(jwk)

        # Verify the proof signature using the embedded JWK
        try:
            from jwt import PyJWK
            key = PyJWK.from_dict(jwk)
            claims = jwt.decode(
                dpop_proof,
                key.key,
                algorithms=['ES256', 'RS256', 'EdDSA'],
                options={"require": ["jti", "htm", "htu", "iat"]}
            )
        except jwt.InvalidTokenError as e:
            raise StrongDMAuthError(f"Invalid DPoP proof: {e}")

        # Verify the HTTP method and URI
        if claims.get('htm') != request.method:
            raise StrongDMAuthError("DPoP proof htm mismatch")

        # Reconstruct the expected URI
        expected_htu = request.url.split('?')[0]  # Remove query string
        if claims.get('htu') != expected_htu:
            raise StrongDMAuthError("DPoP proof htu mismatch")

        # Verify the access token hash (ath) if present
        if 'ath' in claims:
            expected_ath = base64.urlsafe_b64encode(
                hashlib.sha256(access_token.encode()).digest()
            ).rstrip(b'=').decode()
            if claims['ath'] != expected_ath:
                raise StrongDMAuthError("DPoP proof ath mismatch")

        return claims, proof_jkt

    def _verify_token(self, token: str, token_type: str) -> dict:
        """
        Verify a JWT token.

        Args:
            token: The JWT token string
            token_type: 'bearer' or 'dpop'

        Returns:
            The decoded token claims
        """
        try:
            # Peek at unverified claims to determine the issuer and JWKS endpoint
            unverified = jwt.decode(token, options={"verify_signature": False}, algorithms=['RS256', 'ES256', 'EdDSA'])
            token_issuer = unverified.get('iss', '')
            self._validate_issuer(token_issuer)

            # Get the signing key from the token's issuer's JWKS endpoint
            jwks_client = self._get_jwks_client_for_issuer(token_issuer)
            signing_key = jwks_client.get_signing_key_from_jwt(token)

            # Decode and verify the token
            options = {
                "require": ["exp", "iat", "sub"],
                "verify_exp": True,
                "verify_iat": True,
            }

            claims = jwt.decode(
                token,
                signing_key.key,
                algorithms=['RS256', 'ES256', 'EdDSA'],
                audience=self.audience,
                options=options,
            )

            # For DPoP tokens, verify the cnf/jkt claim matches the proof
            if token_type == 'dpop':
                dpop_proof = request.headers.get('DPoP')
                if not dpop_proof:
                    raise StrongDMAuthError("DPoP token requires DPoP proof header")

                # Verify the proof and bind it to token cnf.jkt.
                _, proof_jkt = self._verify_dpop_proof(dpop_proof, token)

                cnf = claims.get('cnf', {})
                token_jkt = cnf.get('jkt')
                if not token_jkt:
                    raise StrongDMAuthError("DPoP token missing cnf.jkt")
                if token_jkt != proof_jkt:
                    raise StrongDMAuthError(
                        "DPoP proof key thumbprint does not match token cnf.jkt"
                    )

            return claims

        except PyJWKClientError as e:
            raise StrongDMAuthError(f"Failed to fetch signing key: {e}")
        except jwt.ExpiredSignatureError:
            raise StrongDMAuthError("Token has expired")
        except jwt.InvalidAudienceError:
            raise StrongDMAuthError("Invalid token audience")
        except jwt.InvalidTokenError as e:
            raise StrongDMAuthError(f"Invalid token: {e}")

    def _introspect_token(self, token: str) -> dict:
        """
        Introspect a token using the introspection endpoint.

        Args:
            token: The token to introspect

        Returns:
            The introspection response
        """
        # Check cache first
        cache_key = hashlib.sha256(token.encode()).hexdigest()[:16]
        if cache_key in self._introspection_cache:
            return self._introspection_cache[cache_key]

        if not self.client_id or not self.client_secret:
            raise StrongDMAuthError(
                "Introspection requires client credentials",
                status_code=500
            )

        introspection_url = f"{self.issuer}/introspect"

        resp = requests.post(
            introspection_url,
            auth=(self.client_id, self.client_secret),
            data={"token": token},
            timeout=10,
        )

        if resp.status_code != 200:
            raise StrongDMAuthError("Token introspection failed")

        result = resp.json()

        # Cache the result
        self._introspection_cache[cache_key] = result

        return result

    def verify_request(self) -> dict:
        """
        Verify the current request's authentication.

        Returns:
            The decoded token claims

        Raises:
            StrongDMAuthError: If authentication fails
        """
        token_type, token = self._extract_token()

        # Try JWT verification first
        try:
            claims = self._verify_token(token, token_type)
            return claims
        except StrongDMAuthError:
            # Never allow DPoP proof failures to bypass sender-constrained validation.
            if not self.introspection_enabled or token_type == 'dpop':
                raise

        # Fall back to introspection
        result = self._introspect_token(token)

        if not result.get('active', False):
            raise StrongDMAuthError("Token is not active")

        return result

    def require_auth(self) -> Callable:
        """
        Decorator to require authentication for a route.

        Usage:
            @app.route('/protected')
            @auth.require_auth()
            def protected():
                return f"Hello, {g.token_claims['sub']}"
        """
        def decorator(f: Callable) -> Callable:
            @wraps(f)
            def decorated_function(*args, **kwargs):
                claims = self.verify_request()
                g.token_claims = claims
                return f(*args, **kwargs)
            return decorated_function
        return decorator

    def require_scope(self, *required_scopes: str, require_all: bool = False) -> Callable:
        """
        Decorator to require specific scopes.

        Args:
            required_scopes: The scopes required (at least one or all, depending on require_all)
            require_all: If True, all scopes are required. If False, at least one is required.

        Usage:
            @app.route('/admin')
            @auth.require_scope('admin')
            def admin():
                return "Admin area"

            @app.route('/restricted')
            @auth.require_scope('admin', require_all=True)
            def restricted():
                return "Restricted area"
        """
        def decorator(f: Callable) -> Callable:
            @wraps(f)
            def decorated_function(*args, **kwargs):
                claims = self.verify_request()
                g.token_claims = claims

                # Get scopes from token
                token_scopes = set(claims.get('scope', '').split())
                required = set(required_scopes)

                if require_all:
                    if not required.issubset(token_scopes):
                        missing = required - token_scopes
                        raise StrongDMAuthError(
                            f"Missing required scopes: {', '.join(missing)}",
                            status_code=403
                        )
                else:
                    if not required.intersection(token_scopes):
                        raise StrongDMAuthError(
                            f"Requires one of: {', '.join(required)}",
                            status_code=403
                        )

                return f(*args, **kwargs)
            return decorated_function
        return decorator

    def get_current_agent(self) -> Optional[dict]:
        """
        Get information about the currently authenticated agent.

        Returns:
            Dict with agent info or None if not authenticated
        """
        claims = getattr(g, 'token_claims', None)
        if not claims:
            return None

        return {
            'subject': claims.get('sub'),
            'scopes': claims.get('scope', '').split(),
            'issuer': claims.get('iss'),
            'client_id': claims.get('client_id') or claims.get('azp'),
            'actor': claims.get('act'),  # For delegated tokens
        }

File: examples/nextjs-middleware/lib/strongdm-auth.ts
/**
 * StrongDM ID Authentication Library for Next.js
 *
 * This library validates JWT tokens issued by StrongDM ID (https://id.strongdm.ai).
 * It supports:
 * - JWT signature verification using JWKS
 * - Token introspection fallback
 * - Scope-based access control
 * - DPoP token validation
 */

import * as jose from "jose";

export interface StrongDMAuthConfig {
  /** Token issuer URL (default: https://id.strongdm.ai) */
  issuer?: string;
  /** Expected audience claim (optional) */
  audience?: string;
  /** Enable token introspection fallback */
  introspectionEnabled?: boolean;
  /** Client ID for introspection */
  clientId?: string;
  /** Client secret for introspection */
  clientSecret?: string;
}

export interface TokenClaims {
  sub: string;
  iss: string;
  iat: number;
  exp: number;
  scope?: string;
  client_id?: string;
  azp?: string;
  act?: {
    sub: string;
  };
  cnf?: {
    jkt?: string;
  };
  [key: string]: unknown;
}

export interface AgentInfo {
  subject: string;
  scopes: string[];
  issuer: string;
  clientId?: string;
  actor?: {
    sub: string;
  };
}

export class StrongDMAuthError extends Error {
  constructor(
    message: string,
    public statusCode: number = 401
  ) {
    super(message);
    this.name = "StrongDMAuthError";
  }
}

// JWKS cache
let jwksCache: jose.JWTVerifyGetKey | null = null;
let jwksCacheExpiry = 0;
const JWKS_CACHE_TTL = 15 * 60 * 1000; // 15 minutes

// Introspection cache
const introspectionCache = new Map<
  string,
  { result: TokenClaims; expiry: number }
>();
const INTROSPECTION_CACHE_TTL = 60 * 1000; // 1 minute

export class StrongDMAuth {
  private config: Required<
    Pick<StrongDMAuthConfig, "issuer"> &
      Omit<StrongDMAuthConfig, "issuer">
  >;

  constructor(config: StrongDMAuthConfig = {}) {
    this.config = {
      issuer: config.issuer || "https://id.strongdm.ai",
      audience: config.audience,
      introspectionEnabled: config.introspectionEnabled || false,
      clientId: config.clientId,
      clientSecret: config.clientSecret,
    };
  }

  /**
   * Get the JWKS keyset for token verification.
   */
  private async getJWKS(): Promise<jose.JWTVerifyGetKey> {
    const now = Date.now();

    if (jwksCache && now < jwksCacheExpiry) {
      return jwksCache;
    }

    const jwksUrl = new URL("/jwks", this.config.issuer);
    jwksCache = jose.createRemoteJWKSet(jwksUrl);
    jwksCacheExpiry = now + JWKS_CACHE_TTL;

    return jwksCache;
  }

  /**
   * Extract the token from the Authorization header.
   */
  extractToken(authHeader: string | null): { type: string; token: string } {
    if (!authHeader) {
      throw new StrongDMAuthError("Missing Authorization header");
    }

    const parts = authHeader.split(" ");
    if (parts.length !== 2) {
      throw new StrongDMAuthError("Invalid Authorization header format");
    }

    const [type, token] = parts;
    const tokenType = type.toLowerCase();

    if (tokenType !== "bearer" && tokenType !== "dpop") {
      throw new StrongDMAuthError(`Unsupported token type: ${type}`);
    }

    return { type: tokenType, token };
  }

  /**
   * Verify a DPoP proof JWT.
   */
  async verifyDPoPProof(
    dpopProof: string,
    accessToken: string,
    method: string,
    url: string
  ): Promise<jose.JWTPayload> {
    // Decode header to get JWK
    const protectedHeader = jose.decodeProtectedHeader(dpopProof);

    if (protectedHeader.typ !== "dpop+jwt") {
      throw new StrongDMAuthError("Invalid DPoP proof: wrong typ");
    }

    if (!protectedHeader.jwk) {
      throw new StrongDMAuthError("Invalid DPoP proof: missing jwk");
    }

    // Import the JWK from the proof header
    const key = await jose.importJWK(protectedHeader.jwk as jose.JWK);

    // Verify the proof
    const { payload } = await jose.jwtVerify(dpopProof, key, {
      typ: "dpop+jwt",
    });

    // Verify required claims
    if (!payload.jti || !payload.htm || !payload.htu || !payload.iat) {
      throw new StrongDMAuthError("Invalid DPoP proof: missing required claims");
    }

    // Verify HTTP method
    if (payload.htm !== method) {
      throw new StrongDMAuthError("DPoP proof htm mismatch");
    }

    // Verify URI (without query string)
    const expectedHtu = url.split("?")[0];
    if (payload.htu !== expectedHtu) {
      throw new StrongDMAuthError("DPoP proof htu mismatch");
    }

    // Verify access token hash if present
    if (payload.ath) {
      const encoder = new TextEncoder();
      const data = encoder.encode(accessToken);
      const hashBuffer = await crypto.subtle.digest("SHA-256", data);
      const hashArray = new Uint8Array(hashBuffer);
      const expectedAth = jose.base64url.encode(hashArray);

      if (payload.ath !== expectedAth) {
        throw new StrongDMAuthError("DPoP proof ath mismatch");
      }
    }

    return payload;
  }

  /**
   * Verify a JWT token.
   */
  async verifyToken(
    token: string,
    tokenType: string,
    dpopProof?: string | null,
    method?: string,
    url?: string
  ): Promise<TokenClaims> {
    const jwks = await this.getJWKS();

    try {
      const { payload } = await jose.jwtVerify(token, jwks, {
        issuer: this.config.issuer,
        audience: this.config.audience,
      });

      // For DPoP tokens, verify the proof
      if (tokenType === "dpop") {
        if (!dpopProof) {
          throw new StrongDMAuthError("DPoP token requires DPoP proof header");
        }
        if (!method || !url) {
          throw new StrongDMAuthError(
            "DPoP verification requires method and url"
          );
        }

        await this.verifyDPoPProof(dpopProof, token, method, url);
      }

      return payload as TokenClaims;
    } catch (error) {
      if (error instanceof StrongDMAuthError) {
        throw error;
      }
      if (error instanceof jose.errors.JWTExpired) {
        throw new StrongDMAuthError("Token has expired");
      }
      if (error instanceof jose.errors.JWTClaimValidationFailed) {
        throw new StrongDMAuthError(`Token validation failed: ${error.message}`);
      }
      throw new StrongDMAuthError(`Invalid token: ${error}`);
    }
  }

  /**
   * Introspect a token using the introspection endpoint.
   */
  async introspectToken(token: string): Promise<TokenClaims> {
    // Check cache
    const cacheKey = await this.hashToken(token);
    const cached = introspectionCache.get(cacheKey);
    if (cached && Date.now() < cached.expiry) {
      return cached.result;
    }

    if (!this.config.clientId || !this.config.clientSecret) {
      throw new StrongDMAuthError(
        "Introspection requires client credentials",
        500
      );
    }

    const introspectionUrl = new URL("/introspect", this.config.issuer);
    const credentials = btoa(
      `${this.config.clientId}:${this.config.clientSecret}`
    );

    const response = await fetch(introspectionUrl.toString(), {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        Authorization: `Basic ${credentials}`,
      },
      body: new URLSearchParams({ token }),
    });

    if (!response.ok) {
      throw new StrongDMAuthError("Token introspection failed");
    }

    const result = await response.json();

    if (!result.active) {
      throw new StrongDMAuthError("Token is not active");
    }

    // Cache the result
    introspectionCache.set(cacheKey, {
      result,
      expiry: Date.now() + INTROSPECTION_CACHE_TTL,
    });

    return result as TokenClaims;
  }

  /**
   * Hash a token for cache key.
   */
  private async hashToken(token: string): Promise<string> {
    const encoder = new TextEncoder();
    const data = encoder.encode(token);
    const hashBuffer = await crypto.subtle.digest("SHA-256", data);
    const hashArray = new Uint8Array(hashBuffer);
    return jose.base64url.encode(hashArray).slice(0, 16);
  }

  /**
   * Verify authentication from request headers.
   */
  async verifyRequest(
    authHeader: string | null,
    dpopHeader?: string | null,
    method?: string,
    url?: string
  ): Promise<TokenClaims> {
    const { type, token } = this.extractToken(authHeader);

    // Try JWT verification first
    try {
      return await this.verifyToken(token, type, dpopHeader, method, url);
    } catch (error) {
      if (!this.config.introspectionEnabled) {
        throw error;
      }
    }

    // Fall back to introspection
    return await this.introspectToken(token);
  }

  /**
   * Check if claims have the required scopes.
   */
  checkScopes(
    claims: TokenClaims,
    requiredScopes: string[],
    requireAll: boolean = false
  ): void {
    const tokenScopes = new Set((claims.scope || "").split(" ").filter(Boolean));
    const required = new Set(requiredScopes);

    if (requireAll) {
      for (const scope of required) {
        if (!tokenScopes.has(scope)) {
          const missing = [...required].filter((s) => !tokenScopes.has(s));
          throw new StrongDMAuthError(
            `Missing required scopes: ${missing.join(", ")}`,
            403
          );
        }
      }
    } else {
      const hasAny = [...required].some((s) => tokenScopes.has(s));
      if (!hasAny) {
        throw new StrongDMAuthError(
          `Requires one of: ${[...required].join(", ")}`,
          403
        );
      }
    }
  }

  /**
   * Get agent info from token claims.
   */
  getAgentInfo(claims: TokenClaims): AgentInfo {
    return {
      subject: claims.sub,
      scopes: (claims.scope || "").split(" ").filter(Boolean),
      issuer: claims.iss,
      clientId: claims.client_id || claims.azp,
      actor: claims.act,
    };
  }
}

/**
 * Create a StrongDM auth instance with environment variables.
 */
export function createStrongDMAuth(): StrongDMAuth {
  return new StrongDMAuth({
    issuer: process.env.STRONGDM_ISSUER || "https://id.strongdm.ai",
    audience: process.env.STRONGDM_AUDIENCE,
    introspectionEnabled: process.env.STRONGDM_INTROSPECTION_ENABLED === "true",
    clientId: process.env.STRONGDM_CLIENT_ID,
    clientSecret: process.env.STRONGDM_CLIENT_SECRET,
  });
}
